### PR TITLE
Add freeze interface to entry and result handlers.

### PR DESCRIPTION
**Pull request by dfish3r**
Update mutable methods to be final and use assertMutable. Refactor heirarchy of handlers that don't need MessageFunctional. Add newInstance to MessageFunctional so that handlers can be mutated by the operation handle. Add convenience constructors and builders to some handlers. Add additional unit tests.

### CODE REVIEW

File: core/src/main/java/org/ldaptive/handler/CaseChangeEntryHandler.java [hunk 1]
```diff
@@ -245,4 +267,109 @@ public String toString()
...
+     */
+    private Builder() {}
+
+
+    /**
+     * Makes this instance immutable.
+     *
+     * @return  this builder
+     */
+    public Builder freeze()
```
Comment (serac) [inline core/src/main/java/org/ldaptive/handler/CaseChangeEntryHandler.java:302]: It strikes me that the object should be implicitly frozen upon calling `build()`, which would

Comment (serac) [inline core/src/main/java/org/ldaptive/handler/CaseChangeEntryHandler.java:302]: It strikes me that the object should be implicitly frozen upon calling `build()`, which would remove the need for this public method.

File: core/src/main/java/org/ldaptive/handler/DnAttributeEntryHandler.java [hunk 2]
```diff
@@ -118,4 +137,81 @@ public String toString()
...
+     */
+    private Builder() {}
+
+
+    /**
+     * Makes this instance immutable.
+     *
+     * @return  this builder
+     */
+    public Builder freeze()
```
Comment (serac) [inline core/src/main/java/org/ldaptive/handler/DnAttributeEntryHandler.java:172]: Same concern as above.

File: core/src/main/java/org/ldaptive/handler/MergeAttributeEntryHandler.java [hunk 3]
```diff
@@ -133,4 +152,81 @@ public String toString()
...
+     */
+    private Builder() {}
+
+
+    /**
+     * Makes this instance immutable.
+     *
+     * @return  this builder
+     */
+    public Builder freeze()
```
Comment (serac) [inline core/src/main/java/org/ldaptive/handler/MergeAttributeEntryHandler.java:187]: Same concern as above.

File: core/src/main/java/org/ldaptive/transport/MessageFunctional.java [hunk 4]
```diff
@@ -1,21 +1,22 @@
...
+ * Base class for processing a message that is initialized with properties from the handle.
  *
  * @param  <Q> type of request
  * @param  <S> type of result
  *
  * @author  Middleware Services
  */
 // CheckStyle:AbstractClassName OFF
-public abstract class MessageFunctional<Q extends Request, S extends Result>
+public abstract class MessageFunctional<Q extends Request, S extends Result> extends AbstractFreezable
```
Comment (serac) [inline core/src/main/java/org/ldaptive/transport/MessageFunctional.java:19]: You have a bit of a naming issue here with an abstract base class that doesn't begin with the typical `Abstract` name. May I suggest `AbstractMessageBasedResultHandler`.

**Review by serac — APPROVED** [review]
Looks good. Had a couple of minor suggestions.

File: core/src/main/java/org/ldaptive/transport/MessageFunctional.java (hunk 4 above)
Comment (dfish3r) [reply core/src/main/java/org/ldaptive/transport/MessageFunctional.java:19]: I'll take that under advisement for a future patch.

**Pull request merged**